### PR TITLE
Bachelor project siller

### PR DESCRIPTION
**Pull request by Gotsiller**
bachelor thesis, validator.py, validator makefile target

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -138,8 +138,13 @@ boolector: smt
 btormc: mc
 	$(foreach file, $(btor2s), btormc $(file);)
 
+# Test testimony for all C* files in sysmbolic
+testimony: tools/Testimony.py
+	chmod 755 tools/Testimony.py
```
Comment (ckirsch) [inline Makefile:143]: This should be unnecessary if you check in the file with proper flags.

File: Makefile [hunk 2]
```diff
@@ -138,8 +138,13 @@ boolector: smt
...
+	$(foreach file, $(wildcard symbolic/*.c), ./tools/Testimony.py $(file);)
+
 # Run everything
-everything: all assemble spike qemu boolector btormc grader grade
+everything: all assemble spike qemu boolector btormc grader grade testimony
```
Comment (ckirsch) [inline Makefile:147]: You need a better name for your tool. How about `validator`?

File: symbolic/division-by-zero-3-35.c [hunk 3]
```diff
@@ -14,7 +14,7 @@ uint64_t main() {
 
   x = malloc(8);
 
-  read(1, x, 1);
+  read(0, x, 1);
```
Comment (ckirsch) [inline symbolic/division-by-zero-3-35.c:17]: Please check with @ChrisEdel if this is ok.

File: theses/README.md [hunk 4]
```diff
@@ -7,4 +7,9 @@ In a world dependent on software, it is of the utmost importance to ensure the c
...
+
+In this bachelor thesis, a parser for Boolector’s witness formatis introduced, which enables to do automated input validation. The software presented takes a C* file containing read calls as input and generates a Btor2 model of it, using Selfie’s model generator. The resulting model is then fed to the Boolector based bound model checker BtorMC. By using a combination of symbolic execution and SMT-solving, BtorMC is able to find error states in the model within a maximum number of executed instructions. In case an error state is found, BtorMC generates a witness. This witness needs to be parsed in order to extract the input string that causes the C* program to run into the error. To validate that the extracted string indeed triggers an error, it is fed to the C* program while running it on Mipster. If the input string was generated correctly, Mipster runs into the predicted error. In addition to the parser, this thesis provides background information on all technologies used and the concepts they are based on.
```
Comment (ckirsch) [inline theses/README.md:15]: This needs spell-checking.

File: symbolic/division-by-zero-3-35.c (hunk 3 above)
Comment (ChrisEdel) [reply symbolic/division-by-zero-3-35.c:17]: Seems fine.

Comment (ckirsch) [reply symbolic/division-by-zero-3-35.c:17]: Why did you use 0?

Comment (ckirsch) [reply symbolic/division-by-zero-3-35.c:17]: What's the difference actually between 0 and 1?

File: Makefile (hunk 2 above)
Comment (Gotsiller) [reply Makefile]: Done.

File: symbolic/division-by-zero-3-35.c (hunk 3 above)
Comment (Gotsiller) [reply symbolic/division-by-zero-3-35.c:17]: 0 = stdin; 1 = stdout; 2 = stderror

**Pull request merged**